### ISSUE TITLE
Discriminated union assignment won't compile inside function

### ISSUE THREAD

**Issue by explosivose** [Microsoft/TypeScript#26258]
**TypeScript Version:**  3.1.0-dev.20180807

**Search Terms:** union type, union type as function parameter, union type with string literal, discriminated union

**Code**

```ts
interface Iface1A {
  type: "A";
}
interface Iface1B {
  type: "B";
}
interface Iface1C {
  type: "C";
}
type Iface1 = Iface1A | Iface1B | Iface1C;

interface Iface2A {
  type: "A";
}
interface Iface2B {
  type: "B";
}
interface Iface2C {
  type: "C";
}
type Iface2 = Iface2A | Iface2B | Iface2C;

const obj1: Iface1 = {type: "C"}; // works
const obj2: Iface2 = {type: obj1.type}; // works
const func = (o1: Iface1) => {
  const o2: Iface2 = {type: o1.type}; // compile error
  // a workaround:
  const o2: Iface2 = {type: o1.type as "A"}; 
  return o2;
};
```

**Expected behavior:** No compile error - should be able to initialize `o2` with the member of `o1` as with `obj1` and `obj2`.

**Actual behavior:** 
```
index.ts:26:9 - error TS2322: Type '{ type: "A" | "B" | "C"; }' is not assignable to type 'Iface2'.
  Type '{ type: "A" | "B" | "C"; }' is not assignable to type 'Iface2C'.
    Types of property 'type' are incompatible.
      Type '"A" | "B" | "C"' is not assignable to type '"C"'.
        Type '"A"' is not assignable to type '"C"'.
```

**Playground Link:** https://www.typescriptlang.org/play/index.html#src=interface%20Iface1A%20%7B%0D%0A%20%20type%3A%20%22A%22%3B%0D%0A%7D%0D%0Ainterface%20Iface1B%20%7B%0D%0A%20%20type%3A%20%22B%22%3B%0D%0A%7D%0D%0Ainterface%20Iface1C%20%7B%0D%0A%20%20type%3A%20%22C%22%3B%0D%0A%7D%0D%0Atype%20Iface1%20%3D%20Iface1A%20%7C%20Iface1B%20%7C%20Iface1C%3B%0D%0A%0D%0Ainterface%20Iface2A%20%7B%0D%0A%20%20type%3A%20%22A%22%3B%0D%0A%7D%0D%0Ainterface%20Iface2B%20%7B%0D%0A%20%20type%3A%20%22B%22%3B%0D%0A%7D%0D%0Ainterface%20Iface2C%20%7B%0D%0A%20%20type%3A%20%22C%22%3B%0D%0A%7D%0D%0Atype%20Iface2%20%3D%20Iface2A%20%7C%20Iface2B%20%7C%20Iface2C%3B%0D%0A%0D%0Aconst%20obj1%3A%20Iface1%20%3D%20%7Btype%3A%20%22C%22%7D%3B%20%2F%2F%20works%0D%0Aconst%20obj2%3A%20Iface2%20%3D%20%7Btype%3A%20obj1.type%7D%3B%20%2F%2F%20works%0D%0Aconst%20func%20%3D%20(o1%3A%20Iface1)%20%3D%3E%20%7B%0D%0A%20%20const%20o2%3A%20Iface2%20%3D%20%7Btype%3A%20o1.type%7D%3B%20%2F%2F%20compile%20error%0D%0A%20%20return%20o2%3B%0D%0A%7D%3B

**Related Issues:**

**Comment by andy-ms** [comment]
So basically:
```ts
type X = { type: "A" } | { type: "B" };
declare const b: boolean;
const x: X = { type: b ? "A" : "B" }; // Error
```

`{ type: "A" | "B" }` isn't considered assignable to `{ type: "A" } | { type: "B" }` currently. I think we've already seen an issue for this but I can't find it.
An easy workaround would be to just define the interface with `type: "A" | "B" | "C"`.

**Comment by RyanCavanaugh** [comment]
Duplicate #12052

**Comment by explosivose** [comment]
Does the workaround apply to discriminated unions?

**Comment by andy-ms** [comment]
If you have a discriminated union like `{ type: "n", n: number } | { type: "s", s: string }`, you can't assign `{ type: "n" | "s" }` to it anyway.

**Comment by RyanCavanaugh** [comment]
If you really have a discriminated union (i.e. the `type` member tells you which other properties are present), then the assignment doesn't make sense to begin with.

**Comment by explosivose** [comment]
I have two different but related discriminated unions which share the set of possible type values. One union is being transformed into the other by a function, so I'd like to simply assign the type and derive the other properties in other ways

**Comment by andy-ms** [comment]
@explosivose Could you show an example? (Obviously there will be type errors if the ways you derive the other properties can't be checked by the compiler.)

**Comment by explosivose** [comment]
Okay this is much closer to the real situation
https://www.typescriptlang.org/play/index.html#src=interface%20ActionTable%20%7B%0D%0A%20%20type%3A%20%22action%22%3B%0D%0A%20%20stage%3A%20number%3B%0D%0A%20%20address%3A%20number%3B%0D%0A%20%20action_data%3A%20%7B%0D%0A%20%20%20%20action_name%3A%20string%3B%0D%0A%20%20%20%20action_args%3A%20string%5B%5D%3B%0D%0A%20%20%7D%3B%0D%0A%20%20action_extra%3A%20%7B%7D%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20MatchTable%20%7B%0D%0A%20%20type%3A%20%22match%22%3B%0D%0A%20%20stage%3A%20number%3B%0D%0A%20%20address%3A%20number%3B%0D%0A%20%20match_type%3A%20string%3B%0D%0A%20%20match_keys%3A%20string%5B%5D%3B%0D%0A%20%20actions%3A%20string%5B%5D%3B%0D%0A%20%20match_extra%3A%20%7B%7D%3B%0D%0A%7D%0D%0Atype%20Table%20%3D%20ActionTable%20%7C%20MatchTable%3B%0D%0A%0D%0Ainterface%20TableSummaryBase%20%7B%0D%0A%20%20stage%3A%20number%3B%0D%0A%20%20address%3A%20number%3B%0D%0A%20%20actions%3A%20string%5B%5D%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20ActionTableSummary%20extends%20TableSummaryBase%20%7B%0D%0A%20%20type%3A%20%22action%22%3B%0D%0A%7D%0D%0Ainterface%20MatchTableSummary%20extends%20TableSummaryBase%20%7B%0D%0A%20%20type%3A%20%22match%22%3B%0D%0A%7D%0D%0Atype%20TableSummary%20%3D%20ActionTableSummary%20%7C%20MatchTableSummary%3B%0D%0A%0D%0Aconst%20func%20%3D%20(table%3A%20Table)%20%3D%3E%20%7B%0D%0A%20%20const%20tableSummary%3A%20TableSummary%20%3D%20%7B%0D%0A%20%20%20%20type%3A%20table.type%2C%0D%0A%20%20%20%20stage%3A%20table.stage%2C%0D%0A%20%20%20%20address%3A%20table.address%2C%0D%0A%20%20%20%20actions%3A%20%5B%5D%2C%0D%0A%20%20%7D%3B%0D%0A%20%20if%20(table.type%20%3D%3D%3D%20%22match%22)%20%7B%0D%0A%20%20%20%20%2F%2F%20derive%20data%20from%20match%20table%0D%0A%20%20%7D%20else%20if%20(table.type%20%3D%3D%3D%20%22action%22)%20%7B%0D%0A%20%20%20%20%2F%2F%20derive%20data%20from%20action%20table%0D%0A%20%20%7D%0D%0A%7D%0D%0A

**Comment by andy-ms** [comment]
That case would also be fixed by combining the interfaces to: `interface TableSummary extends TableSummaryBase { type: "action" | "match"; }`.

**Comment by explosivose** [comment]
Okay thanks for your help guys! I've changed my interfaces 

p.s. the typescript playground is really neat
